dbms/src/Operators/Operator.h

```diff
@@ -0,0 +1,133 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Core/Block.h>
+
+#include <memory>
+
+namespace DB
+{
+/**
+ * All interfaces of the operator may return the following state.
+ * - finish status, spilling status and waiting status can be returned in all interfaces of operator.
+ * - operator may return a different running status depending on the interface.
+*/
+enum class OperatorStatus
+{
+    /// finish status
+    FINISHED,
+    CANCELLED,
+    /// spilling status
+    SPILLING,
+    /// waiting status
+    WAITING,
+    /// running status
+    // means that TransformOp/SinkOp needs to input a block to do the calculation,
+    NEED_INPUT,
+    // means that SourceOp/TransformOp outputs a block as input to the subsequent operators.
+    HAS_OUTPUT,
+};
+
+// TODO support operator profile info like `BlockStreamProfileInfo`.
+
+class Operator
+{
+public:
+    virtual ~Operator() = default;
+    // running status may return are
+    // - `NEED_INPUT` means that the data that the operator is waiting for has been prepared.
+    OperatorStatus await();
+    virtual OperatorStatus awaitImpl() { throw Exception("Unsupport"); }
+    // running status may return are
+    // - `NEED_INPUT` means that operator need data to spill.
+    // - `HAS_OUTPUT` means that operator has restored data, and ready for ouput.
+    OperatorStatus spill();
```

Comment:
Also `SPILLING`, right?

Reply:
Yes, this is just a description of the running status, for finish status, spilling status and waiting status can be returned in any method.